File: engine/parser.js
/**
 * STEP 2 — Parser
 *
 * Converts tokens into:
 * 1) RPN (postfix) using the shunting-yard algorithm
 * 2) An AST (expression tree) from that RPN
 *
 * This module does NOT solve anything. It only structures the equation.
 */

/** @typedef {{ type: 'number', value: number } | { type: 'ident', value: string } | { type: 'op', value: string } | { type: 'paren', value: '('|')'|'['|']' }} Token */

const OPEN_TO_CLOSE = {
  "(": ")",
  "[": "]",
};

const CLOSE_TO_OPEN = {
  ")": "(",
  "]": "[",
};

const PRECEDENCE = {
  "u-": 4,
  "^": 3,
  "*": 2,
  "/": 2,
  "+": 1,
  "-": 1,
};

const RIGHT_ASSOC = new Set(["u-", "^"]);

function isOperatorToken(t) {
  return t.type === "op" && t.value !== "=";
}

function isValueToken(t) {
  return t.type === "number" || t.type === "ident";
}

function isOpenParen(t) {
  return t.type === "paren" && Object.prototype.hasOwnProperty.call(OPEN_TO_CLOSE, t.value);
}

function isCloseParen(t) {
  return t.type === "paren" && Object.prototype.hasOwnProperty.call(CLOSE_TO_OPEN, t.value);
}

function precedenceOf(op) {
  const p = PRECEDENCE[op];
  if (p == null) throw new Error(`Unsupported operator '${op}'.`);
  return p;
}

/**
 * Convert '-' tokens to unary minus ('u-') where appropriate.
 * Unary if '-' occurs:
 *  - at the beginning
 *  - after another operator
 *  - after an opening parenthesis
 *  - after '='
 * @param {Token[]} tokens
 * @returns {Token[]}
 */
function rewriteUnaryMinus(tokens) {
  /** @type {Token[]} */
  const out = [];

  for (let i = 0; i < tokens.length; i++) {
    const t = tokens[i];

    if (t.type === "op" && t.value === "-") {
      const prev = out[out.length - 1];
      const unary =
        !prev ||
        (prev.type === "op" && prev.value !== ")") ||
        (prev.type === "paren" && Object.prototype.hasOwnProperty.call(OPEN_TO_CLOSE, prev.value)) ||
        (prev.type === "op" && prev.value === "=");

      if (unary) {
        out.push({ type: "op", value: "u-" });
        continue;
      }
    }

    out.push(t);
  }

  return out;
}

/**
 * Shunting-yard algorithm: infix tokens -> postfix (RPN).
 * @param {Token[]} tokens Expression tokens ONLY (no '=')
 * @returns {Token[]}
 */
function toRpn(tokens) {
  const input = rewriteUnaryMinus(tokens);

  /** @type {Token[]} */
  const output = [];
  /** @type {Token[]} */
  const ops = [];

  for (const t of input) {
    if (isValueToken(t)) {
      output.push(t);
      continue;
    }

    if (isOpenParen(t)) {
      ops.push(t);
      continue;
    }

    if (isCloseParen(t)) {
      const expectedOpen = CLOSE_TO_OPEN[t.value];
      let found = false;

      while (ops.length) {
        const top = ops.pop();
        if (!top) break;

        if (top.type === "paren" && top.value === expectedOpen) {
          found = true;
          break;
        }

        if (top.type === "op") {
          output.push(top);
        } else {
          throw new Error("Invalid parser state while closing parentheses.");
        }
      }

      if (!found) {
        throw new Error(`Mismatched closing bracket '${t.value}'.`);
      }

      continue;
    }

    if (isOperatorToken(t)) {
      const o1 = t.value;
      const p1 = precedenceOf(o1);

      while (ops.length) {
        const top = ops[ops.length - 1];
        if (!top) break;

        if (top.type === "paren") break;
        if (top.type !== "op") throw new Error("Invalid operator stack state.");

        const o2 = top.value;
        const p2 = precedenceOf(o2);

        const shouldPop = RIGHT_ASSOC.has(o1) ? p1 < p2 : p1 <= p2;
        if (!shouldPop) break;

        output.push(ops.pop());
      }

      ops.push(t);
      continue;
    }

    if (t.type === "op" && t.value === "=") {
      throw new Error("'=' is not allowed inside an expression; split the equation first.");
    }

    throw new Error(`Unsupported token in expression: ${JSON.stringify(t)}`);
  }

  while (ops.length) {
    const top = ops.pop();
    if (!top) break;

    if (top.type === "paren") {
      throw new Error("Mismatched opening bracket.");
    }

    output.push(top);
  }

  return output;
}

/** AST node shapes */

/** @typedef {{ type: 'NumberLiteral', value: number } | { type: 'Variable', name: string } | { type: 'UnaryExpression', op: 'u-', arg: AstNode } | { type: 'BinaryExpression', op: '+'|'-'|'*'|'/', left: AstNode, right: AstNode } AstNode */
/** @typedef {{ type: 'Equation', left: AstNode, right: AstNode }} EquationAst */

function numberNode(value) {
  return { type: "NumberLiteral", value };
}

function variableNode(name) {
  return { type: "Variable", name };
}

function unaryNode(op, arg) {
  return { type: "UnaryExpression", op, arg };
}

function binaryNode(op, left, right) {
  return { type: "BinaryExpression", op, left, right };
}

/**
 * Build an AST from RPN.
 * @param {Token[]} rpn
 * @returns {AstNode}
 */
function rpnToAst(rpn) {
  /** @type {AstNode[]} */
  const stack = [];

  for (const t of rpn) {
    if (t.type === "number") {
      stack.push(numberNode(t.value));
      continue;
    }

    if (t.type === "ident") {
      stack.push(variableNode(t.value));
      continue;
    }

    if (t.type === "op") {
      if (t.value === "u-") {
        const arg = stack.pop();
        if (!arg) throw new Error("Unary '-' missing operand.");
        stack.push(unaryNode("u-", arg));
        continue;
      }

      if (t.value === "+" || t.value === "-" || t.value === "*" || t.value === "/" || t.value === "^") {
        const right = stack.pop();
        const left = stack.pop();
        if (!left || !right) throw new Error(`Operator '${t.value}' missing operand(s).`);
        stack.push(binaryNode(t.value, left, right));
        continue;
      }

      throw new Error(`Unsupported operator in RPN: '${t.value}'.`);
    }

    throw new Error(`Unexpected token in RPN: ${JSON.stringify(t)}`);
  }

  if (stack.length !== 1) {
    throw new Error("Invalid expression: leftover operands/operators after parsing.");
  }

  return stack[0];
}

/**
 * Parse a token list as an expression (no '=' allowed).
 * @param {Token[]} tokens
 * @returns {{ rpn: Token[], ast: AstNode }}
 */
function parseExpression(tokens) {
  const rpn = toRpn(tokens);
  const ast = rpnToAst(rpn);
  return { rpn, ast };
}

/**
 * Split equation tokens around a single '='.
 * @param {Token[]} tokens
 * @returns {{ left: Token[], right: Token[] }}
 */
function splitEquationTokens(tokens) {
  const eqIndexes = [];
  for (let i = 0; i < tokens.length; i++) {
    const t = tokens[i];
    if (t.type === "op" && t.value === "=") eqIndexes.push(i);
  }

  if (eqIndexes.length !== 1) {
    throw new Error("Equation must contain exactly one '='.");
  }

  const idx = eqIndexes[0];
  const left = tokens.slice(0, idx);
  const right = tokens.slice(idx + 1);

  if (!left.length || !right.length) {
    throw new Error("Equation must have expressions on both sides of '='.");
  }

  return { left, right };
}

/**
 * Parse a full equation into { leftAst, rightAst }.
 * @param {Token[]} tokens
 * @returns {{ equation: EquationAst, leftRpn: Token[], rightRpn: Token[] }}
 */
function parseEquation(tokens) {
  const { left, right } = splitEquationTokens(tokens);
  const leftParsed = parseExpression(left);
  const rightParsed = parseExpression(right);

  return {
    equation: { type: "Equation", left: leftParsed.ast, right: rightParsed.ast },
    leftRpn: leftParsed.rpn,
    rightRpn: rightParsed.rpn,
  };
}

/**
 * Pretty-print AST for debugging.
 * @param {AstNode | EquationAst} node
 * @returns {string}
 */
function astToString(node) {
  if (node.type === "Equation") {
    return `${astToString(node.left)} = ${astToString(node.right)}`;
  }

  if (node.type === "NumberLiteral") return String(node.value);
  if (node.type === "Variable") return node.name;

  if (node.type === "UnaryExpression") {
    if (node.op === "u-") return `(-${astToString(node.arg)})`;
    return `(${node.op}${astToString(node.arg)})`;
  }

  if (node.type === "BinaryExpression") {
    return `(${astToString(node.left)} ${node.op} ${astToString(node.right)})`;
  }

  return "<unknown>";
}

module.exports = {
  toRpn,
  rpnToAst,
  parseExpression,
  splitEquationTokens,
  parseEquation,
  astToString,
};

File: engine/rust/src/parser.rs
// ============================================================
//  parser.rs — Port of engine/parser.js
//
//  Two-pass: tokens → RPN (shunting-yard) → AST
// ============================================================

use crate::tokenizer::Token;

// ─── AST ────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub enum AstNode {
    NumberLiteral(i64),
    Variable(char),
    UnaryExpr {
        arg: Box<AstNode>,
    },
    BinaryExpr {
        op: char,
        left: Box<AstNode>,
        right: Box<AstNode>,
    },
}

#[derive(Debug, Clone)]
pub struct EquationAst {
    pub left: AstNode,
    pub right: AstNode,
}

// ─── Helpers ────────────────────────────────────────────────

fn is_open(c: char) -> bool {
    c == '(' || c == '['
}

fn matching_open(close: char) -> char {
    match close {
        ')' => '(',
        ']' => '[',
        _ => panic!("Not a close paren"),
    }
}

fn precedence(op: &str) -> i32 {
    match op {
        "u-" => 3,
        "*" | "/" => 2,
        "+" | "-" => 1,
        _ => panic!("Unknown operator: {}", op),
    }
}

fn is_right_assoc(op: &str) -> bool {
    op == "u-"
}

// Represent operators on the stack as a string (to distinguish 'u-' from '-')
#[derive(Debug, Clone)]
enum StackOp {
    Op(String),
    Paren(char),
}

fn rewrite_unary_minus(tokens: &[Token]) -> Vec<Token> {
    let mut out = Vec::with_capacity(tokens.len());

    for (i, t) in tokens.iter().enumerate() {
        if let Token::Op('-') = t {
            let prev = out.last();
            let unary = match prev {
                None => true,
                Some(Token::Op(c)) if *c != ')' => true,
                Some(Token::Paren(p)) if is_open(*p) => true,
                Some(Token::Op('=')) => true,
                _ => false,
            };
            if unary {
                out.push(Token::UnaryMinus);
                continue;
            }
        }
        out.push(t.clone());
    }
    out
}

fn op_label(t: &Token) -> String {
    match t {
        Token::UnaryMinus => "u-".to_string(),
        Token::Op(c) => c.to_string(),
        _ => panic!("Not an op token"),
    }
}

fn to_rpn(tokens: &[Token]) -> Result<Vec<Token>, String> {
    let input = rewrite_unary_minus(tokens);
    let mut output: Vec<Token> = Vec::new();
    let mut ops: Vec<StackOp> = Vec::new();

    for t in &input {
        match t {
            Token::Number(_) | Token::Ident(_) => {
                output.push(t.clone());
            }

            Token::Paren(c) if is_open(*c) => {
                ops.push(StackOp::Paren(*c));
            }

            Token::Paren(close) => {
                let expected = matching_open(*close);
                let mut found = false;

                while let Some(top) = ops.last() {
                    match top {
                        StackOp::Paren(p) if *p == expected => {
                            ops.pop();
                            found = true;
                            break;
                        }
                        StackOp::Paren(_) => {
                            return Err("Mismatched brackets.".to_string());
                        }
                        StackOp::Op(op) => {
                            let tok = if op == "u-" {
                                Token::UnaryMinus
                            } else {
                                Token::Op(op.chars().next().unwrap())
                            };
                            output.push(tok);
                            ops.pop();
                        }
                    }
                }

                if !found {
                    return Err(format!("Mismatched closing bracket '{}'", close));
                }
            }

            Token::Op('=') => {
                return Err("'=' is not allowed inside an expression.".to_string());
            }

            Token::UnaryMinus | Token::Op(_) => {
                let o1 = op_label(t);
                let p1 = precedence(&o1);

                while let Some(top) = ops.last() {
                    match top {
                        StackOp::Paren(_) => break,
                        StackOp::Op(o2) => {
                            let p2 = precedence(o2);
                            let pop = if is_right_assoc(&o1) { p1 < p2 } else { p1 <= p2 };
                            if !pop {
                                break;
                            }
                            let tok = if o2 == "u-" {
                                Token::UnaryMinus
                            } else {
                                Token::Op(o2.chars().next().unwrap())
                            };
                            output.push(tok);
                            ops.pop();
                        }
                    }
                }

                ops.push(StackOp::Op(o1));
            }

            other => {
                return Err(format!("Unexpected token in expression: {:?}", other));
            }
        }
    }

    while let Some(top) = ops.pop() {
        match top {
            StackOp::Paren(_) => return Err("Mismatched opening bracket.".to_string()),
            StackOp::Op(op) => {
                let tok = if op == "u-" {
                    Token::UnaryMinus
                } else {
                    Token::Op(op.chars().next().unwrap())
                };
                output.push(tok);
            }
        }
    }

    Ok(output)
}

fn rpn_to_ast(rpn: &[Token]) -> Result<AstNode, String> {
    let mut stack: Vec<AstNode> = Vec::new();

    for t in rpn {
        match t {
            Token::Number(n) => stack.push(AstNode::NumberLiteral(*n)),
            Token::Ident(c) => stack.push(AstNode::Variable(*c)),

            Token::UnaryMinus => {
                let arg = stack.pop().ok_or("Unary '-' missing operand.")?;
                stack.push(AstNode::UnaryExpr { arg: Box::new(arg) });
            }

            Token::Op(op) if matches!(*op, '+' | '-' | '*' | '/') => {
                let right = stack.pop().ok_or_else(|| format!("Op '{}' missing right operand.", op))?;
                let left  = stack.pop().ok_or_else(|| format!("Op '{}' missing left operand.",  op))?;
                stack.push(AstNode::BinaryExpr {
                    op: *op,
                    left:  Box::new(left),
                    right: Box::new(right),
                });
            }

            other => return Err(format!("Unexpected token in RPN: {:?}", other)),
        }
    }

    if stack.len() != 1 {
        return Err("Invalid expression: leftover operands after parsing.".to_string());
    }

    Ok(stack.pop().unwrap())
}

fn split_on_equals(tokens: &[Token]) -> Result<(&[Token], &[Token]), String> {
    let indexes: Vec<usize> = tokens
        .iter()
        .enumerate()
        .filter(|(_, t)| matches!(t, Token::Op('=')))
        .map(|(i, _)| i)
        .collect();

    if indexes.len() != 1 {
        return Err("Equation must contain exactly one '='.".to_string());
    }

    let idx = indexes[0];
    let left  = &tokens[..idx];
    let right = &tokens[idx + 1..];

    if left.is_empty() || right.is_empty() {
        return Err("Equation must have expressions on both sides of '='.".to_string());
    }

    Ok((left, right))
}

fn parse_expression(tokens: &[Token]) -> Result<AstNode, String> {
    let rpn = to_rpn(tokens)?;
    rpn_to_ast(&rpn)
}

pub fn parse_equation(tokens: &[Token]) -> Result<EquationAst, String> {
    let (left_toks, right_toks) = split_on_equals(tokens)?;
    let left  = parse_expression(left_toks)?;
    let right = parse_expression(right_toks)?;
    Ok(EquationAst { left, right })
}
